### PR TITLE
Re establish connection on errors.

### PR DESCRIPTION
**Pull request by gsaluja9**
This would handle scenarios where a long running connection would be checked for errors, with logic to recover.
Reported as in #258

### CODE REVIEW

**Comment by gsaluja9** [comment]
This is blocked on this : https://github.com/aperture-data/athena/pull/1205

File: aperturedb/Configuration.py [hunk 1]
```diff
@@ -14,6 +14,10 @@ class Configuration:
     use_ssl: bool = True
     use_rest: bool = False
     use_keepalive: bool = True
+    retry_connect_interval_seconds: int = 10
+    # Max number of attempts to retry the initial connection (0 means infinite)
+    # This is useful when the aperturedb server is not ready yet.
+    retry_connect_max_attempts: int = 0
```
Comment (drewaogle) [inline aperturedb/Configuration.py:20]: lets not do infinite as the default. I think expect a library to return an error at some point.

File: test/test_Session.py [hunk 2]
```diff
@@ -79,3 +82,139 @@ def mock_send_msg(msg):
...
+            new_db = Connector(
+                dbinfo.DB_TCP_HOST,
+                dbinfo.DB_TCP_PORT,
+                dbinfo.DB_USER,
+                dbinfo.DB_PASSWORD,
+                retry_connect_max_attempts=3,
+                retry_connect_interval_seconds=0)
+        except Exception as e:
+            # Check the exception is not an obscure one.
+            assert str(e).startswith("Could not connect to apertureDB server:")
```
Comment (drewaogle) [inline test/test_Session.py:107]: nice test!

**Review by drewaogle — COMMENTED** [review]
just the question about the default otherwise lgtm

File: aperturedb/Configuration.py (hunk 1 above)
Comment (gsaluja9) [reply aperturedb/Configuration.py:20]: I thought so too myself, but was convinced otherwise thinking of attaining analogous behavior and a current customer (watchdog).

1. So ConnectorRest does not encapsulate a connection. Not doing like this would introduce 2 separate default behaviors using rest or not.
2. The watchdog .... with increasing number of URLs, it'll be hard to ensure that we start it after all servers are up, and this is the case that got @luisremis because they anyways did not care (or write code to handle the response after creating the connector. I think this behavior can be explicitly documented but it gives the connectors a parity on how they init.

If you are convinced then, I can update the docs to highlight this, otherwise lets discuss this more.

File: aperturedb/Connector.py [hunk 3]
```diff
@@ -383,8 +425,9 @@ def query(self, q, blobs=[]):
             self.last_query_time = time.time() - start
             return self.response, self.blobs
         except BaseException as e:
-            logger.critical(e)
-            raise ConnectionError("ApertureDB disconnected")
+            logger.critical(traceback.format_exc(limit=10))
```
Comment (bovlb) [inline aperturedb/Connector.py:428]: Why not use `exc_info` here?

File: aperturedb/Connector.py [hunk 4]
```diff
@@ -383,8 +425,9 @@ def query(self, q, blobs=[]):
...
+            logger.critical("\r\n".join(traceback.format_stack(limit=10)))
```
Comment (bovlb) [inline aperturedb/Connector.py:429]: Why not use `stack_info` here?

File: test/docker-compose.yml [hunk 5]
```diff
@@ -2,7 +2,7 @@ version: '3'
 
 services:
   aperturedb:
-    image: aperturedata/aperturedb
+    image: aperturedata/aperturedb:vdevelop
```
Comment (bovlb) [inline test/docker-compose.yml:5]: Why the change in image?

File: aperturedb/Configuration.py (hunk 1 above)
Comment (gsaluja9) [reply aperturedb/Configuration.py:20]: OK, I checked ConnectorRest and I stand corrected. Even that tries to connect in init, and bails out if server is not up.
I shall make this from infinite to 3, and 1 second for retry interval as default. 
Thanks for pointing this out.

File: test/test_Session.py (hunk 2 above)
Comment (gsaluja9) [reply test/test_Session.py:107]: Discussed this in IM.

File: test/docker-compose.yml (hunk 5 above)
Comment (gsaluja9) [reply test/docker-compose.yml:5]: Same as above

Comment (drewaogle) [reply test/docker-compose.yml:5]: cool.

File: aperturedb/Configuration.py (hunk 1 above)
Comment (drewaogle) [reply aperturedb/Configuration.py:20]: good to know!

We should put this in docs so we know our theory of operation on this stuff.

File: aperturedb/Connector.py (hunk 3 above)
Comment (gsaluja9) [reply aperturedb/Connector.py:428]: will address.

File: test/docker-compose.yml (hunk 5 above)
Comment (luisremis) [reply test/docker-compose.yml:5]: [v0.15.16](https://github.com/aperture-data/athena/compare/v0.15.15...v0.15.16) is ready to go!

**Review by luisremis — DISMISSED** [review]
🚀

File: test/docker-compose.yml (hunk 5 above)
Comment (drewaogle) [reply test/docker-compose.yml]: I think we can resolve this
